airflow/jobs.py

```diff
@@ -484,84 +484,99 @@ def process_dag(self, dag, queue):
             return None
-        else:
-            # Taking a lock
-            db_dag.scheduler_lock = True
-            db_dag.last_scheduler_run = datetime.now()
-            session.commit()
 
-        active_runs = dag.get_active_runs()
+        # Release the db lock
+        # the assumption here is that process_dag will take less
+        # time than self.heartrate otherwise we might unlock too
+        # quickly and this should moved below, but that would increase
+        # the time the record is locked and is blocking for other calls.
+        db_dag.last_scheduler_run = datetime.now()
+        session.commit()
 
-        self.logger.info('Getting list of tasks to skip for active runs.')
-        skip_tis = set()
-        if active_runs:
-            qry = (
-                session.query(TI.task_id, TI.execution_date)
-                .filter(
-                    TI.dag_id == dag.dag_id,
-                    TI.execution_date.in_(active_runs),
-                    TI.state.in_((State.RUNNING, State.SUCCESS, State.FAILED)),
-                )
-            )
-            skip_tis = {(ti[0], ti[1]) for ti in qry.all()}
+        # update the state of the previously active dag runs
+        dag_runs = DagRun.find(dag_id=dag.dag_id, state=State.RUNNING)
+        active_dag_runs = []
+        for run in dag_runs:
+            # do not consider runs that are executed in the future
+            if run.execution_date > datetime.now():
+                continue
 
-        descartes = [obj for obj in product(dag.tasks, active_runs)]
-        could_not_run = set()
-        self.logger.info('Checking dependencies on {} tasks instances, minus {} '
-                     'skippable ones'.format(len(descartes), len(skip_tis)))
+            # todo: run.task is transient but needs to be set
+            run.dag = dag
+            # todo: preferably the integrity check happens at dag collection time
+            run.verify_integrity()
+            run.update_state()
+            if run.state == State.RUNNING:
+                active_dag_runs.append(run)
 
-        for task, dttm in descartes:
-            if task.adhoc or (task.task_id, dttm) in skip_tis:
-                continue
-            ti = TI(task, dttm)
+        could_not_run = 0
+        available_tasks = 0
 
-            ti.refresh_from_db()
-            if ti.state in (
-                    State.RUNNING, State.QUEUED, State.SUCCESS, State.FAILED):
-                continue
-            elif ti.is_runnable(flag_upstream_failed=True):
-                self.logger.debug('Queuing task: {}'.format(ti))
-                queue.put((ti.key, pickle_id))
-            elif ti.is_premature():
-                continue
-            else:
-                self.logger.debug('Adding task: {} to the COULD_NOT_RUN set'.format(ti))
-                could_not_run.add(ti)
+        for run in active_dag_runs:
+            tis = run.get_task_instances(session=session, state=(State.NONE,
```

Comment:
I think this line is too restrictive.

As an example, using CeleryExecutor but no workers, this DAG is almost immediately marked as failed (I had it on hand from testing a different issue):
https://gist.github.com/jlowin/4f0c88b80f35ac45862c6222d153abcf

I think what happens is that the first set of tasks are loaded and sent to the Executor, so their states are set to RUNNING even though there are no celery workers. The next time through the loop, the scheduler gets to this line and loads up the one remaining task -- one that is also downstream from all the RUNNING tasks and therefore can't run yet. It evaluates it, finds it can't run, and says the DAG is deadlocked because `could_not_run == available_tasks`. 

So I think we need to account for the fact that tasks may be running or in transit to the executor before moving on to deadlock assumption.